dll/win32/shell32/wine/shellpath.c

```diff
@@ -106,6 +106,127 @@ DoGetProductType(PNT_PRODUCT_TYPE ProductType)
 	########## Combining and Constructing paths ##########
 */
 
+/* @implemented */
+static BOOL WINAPI
+PathSearchOnExtensionsW(LPWSTR path, LPCWSTR *dirs, BOOL flag, DWORD dwWhich)
+{
+    if (*PathFindExtensionW(path) != 0)
+        return FALSE;
+
+    if (flag)
+        return PathFindOnPathExW(path, dirs, dwWhich);
+    else
+        return PathFileExistsDefExtW(path, dwWhich);
+}
+
+/* #define REQUIREABSOLUTE */
+
+#ifdef REQUIREABSOLUTE
+/* @implemented */
+static BOOL WINAPI PathIsAbsoluteW(LPCWSTR path)
+{
+    return PathIsUNCW(path) || (PathGetDriveNumberW(path) != -1 && path[2] == L'\\');
+}
+
+/* @implemented */
+static BOOL WINAPI PathMakeAbsoluteW(LPWSTR path)
+{
+    WCHAR path1[MAX_PATH];
+    DWORD cch;
+
+    if (path == NULL)
+        return FALSE;
+    cch = GetCurrentDirectoryW(MAX_PATH, path1);
+    if (!cch || cch > MAX_PATH)
+        return FALSE;
+    return (PathCombineW(path, path1, path) != NULL);
+}
+#endif /* def REQUIREABSOLUTE */
+
+static DWORD GetShortPathNameAbsentW(LPCWSTR pszLong, LPWSTR pszShort, DWORD cchShort)
+{
+    FIXME("GetShortPathNameAbsentW(%ls, %p, %ld): stub\n", pszLong, pszShort, cchShort);
+    StringCchCopyW(pszShort, cchShort, pszLong);
+    return lstrlenW(pszShort);
+}
+
+BOOL WINAPI IsLFNDriveW(LPCWSTR lpszPath);
+
+/* @unconfirmed */
+static VOID WINAPI PathQualifyExW(LPWSTR pszPath, LPCWSTR pszDir, DWORD dwFlags)
+{
+    WCHAR szRoot[MAX_PATH], szCopy[MAX_PATH], szCurDir[MAX_PATH];
+    LPWSTR pch;
+    LONG cch;
+    BOOL bCheckLFN;
+
+    if (FAILED(StringCchCopyW(szCopy, _countof(szCopy), pszPath)))
+        return;
+
+    FixSlashesAndColonW(szCopy);
+
+    if (pszDir)
+    {
+        cch = GetCurrentDirectoryW(_countof(szCurDir), szCurDir);
+        if (cch <= 0 || cch >= MAX_PATH || !SetCurrentDirectoryW(pszDir))
```

Comment:
What does cch correspond to? Length for pszDir ?

Reply:
It checks the return value of `GetCurrentDirectoryW`.

Reply:
Sure, but then, is the `MAX_PATH` check somehow a condition on what should be the max size of the buffer pointed by pszDir?

Reply:
That's for `szCurDir`.

Reply:
No, I don't.